File: bridge/java/Bridge.java
interface Computer {
    void print();
}

class Mac implements Computer {
    private final Printer printer;

    public Mac(Printer printer) {
        this.printer = printer;
    }

    @Override
    public void print() {
        System.out.println("Print request for Mac");
        printer.printFile();
    }
}

class Linux implements Computer {
    private final Printer printer;

    public Linux(Printer printer) {
        this.printer = printer;
    }

    @Override
    public void print() {
        System.out.println("Print request for Linux");
        printer.printFile();
    }
}

class Windows implements Computer {
    private final Printer printer;

    public Windows(Printer printer) {
        this.printer = printer;
    }

    @Override
    public void print() {
        System.out.println("Print request for Windows");
        printer.printFile();
    }
}

// ============================================================

interface Printer {
    void printFile();
}

class Epson implements Printer {
    @Override
    public void printFile() {
        System.out.println("Printing by a EPSON Printer");
    }
}

class Hp implements Printer {
    @Override
    public void printFile() {
        System.out.println("Printing by a HP Printer");
    }
}

File: bridge/go/bridge.go
package bridge

import "fmt"

type Computer interface {
	print()
}

type Mac struct {
	printer Printer
}

func (m *Mac) print() {
	fmt.Println("Print request for Mac")
	m.printer.printFile()
}

type Linux struct {
	printer Printer
}

func (l *Linux) print() {
	fmt.Println("Print request for Linux")
	l.printer.printFile()
}

type Windows struct {
	printer Printer
}

func (w *Windows) print() {
	fmt.Println("Print request for Windows")
	w.printer.printFile()
}

// ============================================================

type Printer interface {
	printFile()
}

type Epson struct{}

func (*Epson) printFile() {
	fmt.Println("Printing by a EPSON Printer")
}

type Hp struct{}

func (*Hp) printFile() {
	fmt.Println("Printing by a Hp Printer")
}
